cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -466,80 +469,87 @@ ReconciliationState createReconciliationState(Reconciliation reconciliation, Kaf
-            return waitForQuiescence(namespace, kafkaSsName).compose(
-                ss -> {
-                    if (ss == null) {
-                        return Future.succeededFuture(this);
-                    }
-                    log.debug("Does SS {} need to be upgraded?", ss.getMetadata().getName());
-                    Future<?> result;
-                    // Get the current version of the cluster
-                    // Strimzi 0.8.x and 0.9.0 didn't set the annotation, so if it's absent we know it must be 2.0.0
-                    KafkaVersion currentVersion = versions.version(Annotations.annotations(ss).getOrDefault(ANNO_STRIMZI_IO_KAFKA_VERSION, "2.0.0"));
-                    log.debug("SS {} has current version {}", ss.getMetadata().getName(), currentVersion);
-                    String fromVersionAnno = Annotations.annotations(ss).get(ANNO_STRIMZI_IO_FROM_VERSION);
-                    KafkaVersion fromVersion;
-                    if (fromVersionAnno != null) { // We're mid-upgrade
-                        fromVersion = versions.version(fromVersionAnno);
-                    } else {
-                        fromVersion = currentVersion;
-                    }
-                    log.debug("SS {} is from version {}", ss.getMetadata().getName(), fromVersion);
-                    String toVersionAnno = Annotations.annotations(ss).get(ANNO_STRIMZI_IO_TO_VERSION);
-                    KafkaVersion toVersion;
-                    if (toVersionAnno != null) { // We're mid-upgrade
-                        toVersion = versions.version(toVersionAnno);
-                    } else {
-                        toVersion = versions.version(kafkaAssembly.getSpec().getKafka().getVersion());
-                    }
-                    log.debug("SS {} is to version {}", ss.getMetadata().getName(), toVersion);
-                    KafkaUpgrade upgrade = new KafkaUpgrade(fromVersion, toVersion);
-                    log.debug("Kafka upgrade {}", upgrade);
-                    if (upgrade.isNoop()) {
-                        log.debug("Kafka.spec.kafka.version unchanged");
-                        result = Future.succeededFuture();
-                    } else {
-                        String image = versions.kafkaImage(kafkaAssembly.getSpec().getKafka().getImage(), toVersion.version());
-                        Future<StatefulSet> f = Future.succeededFuture(ss);
-                        if (upgrade.isUpgrade()) {
-                            if (currentVersion.equals(fromVersion)) {
-                                f = f.compose(ignored -> kafkaUpgradePhase1(ss, upgrade, image));
-                            }
-                            result = f.compose(ss2 -> kafkaUpgradePhase2(ss2, upgrade));
+            return waitForQuiescence(namespace, kafkaSsName).compose(quiescence -> {
+                if (!quiescence) {
+                    return Future.succeededFuture(this);
```

Comment:
Should this really return success? That will continue the reconciliation which can for example try to reconfigure the broker with new configuration changes without doing the upgrade, or? I think that is dangerous. What does guarantee that the rest of the reconciliation doesn't create any problems?

(At the same time you cannot fail the reconciliation because that would probably never make `quiescence` become `true`.)

Reply:
If the cluster is not "stable" about the generation (related to the "check" discussion), I would say that it's better to avoid the upgrade, or? Is that what you had in mind @tombentley ?

Reply:
I do not think this is the right question. The right question to ask is _What happens if the user changed the `version` in CR from 2.0.0 to 2.1.1 and we instead of doing the upgrade let it go through the regular reconciliation and regular rolling update?_ Will it really just fix the rolling update generation without doing any other changes? Or will it do more than that?

Reply:
Imho it should do the regular rolling update and then doing the upgrade on the next reconciliation.

Reply:
Why do you think it would still use the image with the old version, use the old configuration etc.?

Reply:
@scholzj has a point here, actually. The regular reconciliation will end up replacing the images in the STS, which would put the cluster into a bad state. To make it work we would need to ensure we didn't honour the `version` (or `image`) in the `Kafka` resource during that regular reconciliation. Doing that would make the code quite complicated. So I'm beginning to think that the better way to solve this would be to do as @scholzj originally suggested and simply move the old `waitForQuiessence()` method into the `!noop` branch. Sorry for leading you astray @ppatierno.

Reply:
Thinking about this more... I now think that @ppatierno's original solution (moving `kafkaManualPodCleaning()` and `kafkaManualRollingUpdate()` before the call to `kafkaUpgrade()`) is correct. Neither `kafkaManualPodCleaning()` nor `kafkaManualRollingUpdate()` change the statefulset, so they can delete pods without those pods being restarted with a changed image. That means the user can annotate the STS to trigger the pod cleaning/manual rolling and those will take effect before we enter  (the original version of)`waitForQuiescence()`. If `waitForQuiescence()` times out then we exit the reconciliation loop and a subsequent reconciliation might succeed. Am I missing something?

Reply:
That might be. But we should still do the `waitForQuiescence` only if it is needed. We should not block the reconciliation loop here when no upgrade is going on.

Reply:
I will do then, agree that both cover the issue.